multiversx_sdk_core/transaction_factories/smart_contract_factory.py

```diff
@@ -0,0 +1,192 @@
+                                           TRANSACTION_OPTIONS_DEFAULT,
+                                           TRANSACTION_VERSION_DEFAULT,
+                                           VM_TYPE_WASM_VM)
+from multiversx_sdk_core.interfaces import (IAddress, IChainID, IGasLimit,
+                                            IGasPrice, INonce,
+                                            ITransactionValue)
+from multiversx_sdk_core.serializer import arg_to_string
+from multiversx_sdk_core.transaction import Transaction
+from multiversx_sdk_core.transaction_payload import TransactionPayload
+from multiversx_sdk_core.utils import read_binary_file
+
+
+class IConfig(Protocol):
+    chain_id: IChainID
+
+
+class CodeMetadata:
+    def __init__(self, upgradeable: bool = True, readable: bool = True, payable: bool = False, payable_by_sc: bool = False) -> None:
+        self.upgradeable = upgradeable
+        self.readable = readable
+        self.payable = payable
+        self.payable_by_sc = payable_by_sc
+
+    def to_hex(self) -> str:
+        flag_value_pairs = [
+            (0x01_00, self.upgradeable),
+            (0x04_00, self.readable),
+            (0x00_02, self.payable),
+            (0x00_04, self.payable_by_sc)
+        ]
+        metadata_value = self.sum_flag_values(flag_value_pairs)
+        return f"{metadata_value:04X}"
+
+    def sum_flag_values(self, flag_value_pairs: List[Tuple[int, bool]]) -> int:
+        value_sum = 0
+        for value, flag in flag_value_pairs:
+            if flag:
+                value_sum += value
+        return value_sum
+
+
+class SmartContractFactory:
+    def __init__(self, config: IConfig) -> None:
+        self.config = config
+
+    def deploy(self,
+               deployer: IAddress,
+               nonce: INonce,
+               bytecode_path: Path,
+               gas_limit: IGasLimit,
+               arguments: Optional[List[Any]] = None,
+               value: Optional[ITransactionValue] = None,
+               gas_price: Optional[IGasPrice] = None,
+               is_upgradeable: bool = True,
+               is_readable: bool = True,
+               is_payable: bool = False,
+               is_payable_by_sc: bool = False) -> Transaction:
+        bytecode = read_binary_file(bytecode_path)
+        metadata = CodeMetadata(is_upgradeable, is_readable, is_payable, is_payable_by_sc)
+        arguments = arguments or []
+
+        parts = [
+            arg_to_string(bytecode),
+            arg_to_string(VM_TYPE_WASM_VM),
+            metadata.to_hex()
+        ]
+
+        for arg in arguments:
+            parts.append(arg_to_string(arg))
+
+        transaction = self.create_transaction(
+            sender=deployer,
+            receiver=Address.from_bech32(CONTRACT_DEPLOY_ADDRESS),
+            data_parts=parts,
+            gas_limit=gas_limit,
+            gas_price=gas_price,
+            nonce=nonce,
+            value=value
+        )
+
+        return transaction
```

Comment:
Thinking out loud - perhaps we should also return the smart contract address here? Though, if we do so, we do require the nonce as input, indeed. Or perhaps, return the address as `Optional[Address]` and only if nonce is provided in the input? Or simply do not return the address (as it is now), and let the developer call `compute_contract_address` of `address.py`.

Reply:
I was thinking about returning the contract address as well, but then the nonce should be required. I don't think it's a good idea to return a `Tuple[Transaction, Address]` if the `nonce` it's provided else just return the transaction. I'd say it's pretty easy for the developer to compute the contract address if needed. Will leave it as it is, for now, but we can change it in the future if needed.